File: mvvm_simple/src/main/java/com/example/mvvmdemo/view/MainActivity.java
package com.example.mvvmdemo.view;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;

import com.example.mvvmdemo.databinding.ActivityMainBinding;
import com.example.mvvmdemo.model.UserInfo;
import com.example.mvvmdemo.viewmodel.UserInfoViewModel;

public class MainActivity extends AppCompatActivity {

    private UserInfoViewModel userInfoViewModel;
    private ProgressDialog progressDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityMainBinding viewBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(viewBinding.getRoot());
        userInfoViewModel = new UserInfoViewModel();
        userInfoViewModel.getLoadingLiveData().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean aBoolean) {
                if (aBoolean) {
                    showLoading();
                } else {
                    hideLoading();
                }
            }
        });
        userInfoViewModel.getUserInfoLiveData().observe(this, new Observer<UserInfo>() {
            @Override
            public void onChanged(UserInfo userInfo) {
                viewBinding.tvUserInfo.setText(
                        String.format("姓名：%s，年龄：%d，地址：%s",
                                userInfo.getName(), userInfo.getAge(), userInfo.getAddress())
                );
            }
        });

        viewBinding.btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                userInfoViewModel.getUserInfo();
            }
        });
    }

    public void showLoading() {
        if (progressDialog == null) {
            progressDialog = new ProgressDialog(this);
            progressDialog.setMessage("正在加载");
        }
        progressDialog.show();
    }

    public void hideLoading() {
        progressDialog.hide();
    }
}

File: mvvm_retrofit_coroutines/src/main/java/com/example/myapplication/MainActivity.kt
package com.example.myapplication

import android.app.ProgressDialog
import android.os.Bundle
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import com.example.myapplication.databinding.ActivityMainBinding
import com.example.myapplication.entity.DataResult

class MainActivity : AppCompatActivity() {

    private val viewModel: MainViewModel by viewModels()

    private lateinit var loadingDialog: ProgressDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        loadingDialog = ProgressDialog(this)

        binding.btnRequest.setOnClickListener {
            showLoading()
            viewModel.search1(1, "鸿洋").observe(this) {
                when (it) {
                    is DataResult.Success -> {
                        hideLoading()
                        binding.tvContent.text = it.response.toString()
                    }
                    is DataResult.Error -> {
                        hideLoading()
                        toast(it.exception.msg)
                    }
                }
            }
        }

        binding.btnRequest2.setOnClickListener {
            showLoading()
            viewModel.search2(1, "鸿洋").observe(this) {
                when (it) {
                    is DataResult.Success -> {
                        hideLoading()
                        binding.tvContent.text = it.response.toString()
                    }
                    is DataResult.Error -> {
                        hideLoading()
                        toast(it.exception.msg)
                    }
                }
            }
        }
    }

    private fun toast(msg: String) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }

    private fun showLoading() {
        loadingDialog.show()
    }

    private fun hideLoading() {
        loadingDialog.dismiss()
    }
}